frontendGraph/src/main/java/com/virtuslab/gitmachete/frontend/graph/print/PrintElementGeneratorImpl.java

```diff
@@ -0,0 +1,103 @@
+import com.intellij.vcs.log.graph.api.printer.PrintElementGenerator;
+import com.intellij.vcs.log.graph.impl.print.elements.EdgePrintElementImpl;
+import com.intellij.vcs.log.graph.impl.print.elements.PrintElementWithGraphElement;
+import com.intellij.vcs.log.graph.impl.print.elements.SimplePrintElementImpl;
+import lombok.RequiredArgsConstructor;
+
+import com.virtuslab.gitmachete.frontend.graph.GraphElementManager;
+import com.virtuslab.gitmachete.frontend.graph.elements.BranchElement;
+import com.virtuslab.gitmachete.frontend.graph.elements.IGraphElement;
+import com.virtuslab.gitmachete.frontend.graph.repository.RepositoryGraph;
+
+public final class PrintElementGeneratorImpl implements PrintElementGenerator {
+  private final RepositoryGraph repositoryGraph;
+  private final GraphElementManager printElementManager;
+
+  public PrintElementGeneratorImpl(RepositoryGraph repositoryGraph, GraphElementManager graphElementManager) {
+    this.repositoryGraph = repositoryGraph;
+    this.printElementManager = graphElementManager;
+  }
+
+  @Override
+  public Collection<PrintElementWithGraphElement> getPrintElements(int rowIndex) {
+    PrintElementBuilder builder = new PrintElementBuilder(rowIndex);
+    collectElements(rowIndex, builder);
+    return builder.build();
+  }
+
+  private void collectElements(int rowIndex, PrintElementBuilder builder) {
+    GraphNode graphNode = repositoryGraph.getGraphNode(rowIndex);
+    IGraphElement graphElement = repositoryGraph.getGraphElement(rowIndex);
+    int position = graphElement.getIndentLevel();
+
+    repositoryGraph.getVisibleEdgesWithPositions(rowIndex).forEach(edgeAndPos -> {
+      builder.consumeUpEdge(edgeAndPos._1(), edgeAndPos._2());
+      builder.consumeDownEdge(edgeAndPos._1(), edgeAndPos._2());
+    });
+
+    java.util.List<GraphEdge> adjacentEdges = repositoryGraph.getAdjacentEdges(rowIndex, EdgeFilter.ALL);
+    for (GraphEdge edge : adjacentEdges) {
+      Integer downNodeIndex = edge.getDownNodeIndex();
+      Integer upNodeIndex = edge.getUpNodeIndex();
+      if (downNodeIndex != null && downNodeIndex == rowIndex) {
+        builder.consumeUpEdge(edge, position);
+      }
+      if (upNodeIndex != null && upNodeIndex == rowIndex) {
+        builder.consumeDownEdge(edge, position);
+      }
+    }
+
+    int nodeAndItsDownEdgePos = position;
+    if (graphElement.isBranch() && !((BranchElement) graphElement).getBranch().isRootBranch()) {
+      builder.consumeRightEdge(new GraphEdge(rowIndex, rowIndex, null, GraphEdgeType.USUAL), position);
+      nodeAndItsDownEdgePos++;
+    }
+
+    builder.consumeNode(graphNode, nodeAndItsDownEdgePos);
+    if (graphElement.hasSubelement()) {
+      builder.consumeDownEdge(new GraphEdge(rowIndex, rowIndex + 1, null, GraphEdgeType.USUAL), nodeAndItsDownEdgePos);
+    }
+  }
+
+  @RequiredArgsConstructor
+  private final class PrintElementBuilder {
+    private final ArrayList<PrintElementWithGraphElement> result = new ArrayList<>();
```

Comment:
Nit: `List` in the declared type unless `ArrayList`-specific features are necessary

Reply:
And maybe `SmartList` on the right-hand side?